pkg/fuzzer/job.go

```diff
@@ -0,0 +1,428 @@
+	stop bool) {
+	const signalRuns = 3
+	// TODO: run all signalRuns in parallel?
+	var notExecuted int
+	for i := 0; i < signalRuns; i++ {
+		result := fuzzer.execWait(job, &Request{
+			Prog:         job.p,
+			NeedSignal:   true,
+			NeedCover:    true,
+			NeedRawCover: fuzzer.Config.FetchRawCover,
+			stat:         statTriage,
+		})
+		if result.Stop {
+			stop = true
+			return
+		}
+		info := result.Info
+		if !reexecutionSuccess(info, &job.info, job.call) {
+			// The call was not executed or failed.
+			notExecuted++
+			if notExecuted >= signalRuns/2+1 {
+				return // if happens too often, give up
+			}
+			continue
+		}
+		thisSignal, thisCover := getSignalAndCover(job.p, info, job.call)
+		if len(rawCover) == 0 && fuzzer.Config.FetchRawCover {
+			rawCover = append([]uint32{}, thisCover...)
+		}
+		if stableSignal.Empty() {
+			stableSignal = newSignal.Intersection(thisSignal)
+		} else {
+			stableSignal = stableSignal.Intersection(thisSignal)
+		}
+		if stableSignal.Empty() {
+			return
+		}
+		inputCover.Merge(thisCover)
+	}
+	return
+}
+
+func (job *triageJob) minimize(fuzzer *Fuzzer, newSignal signal.Signal) (stop bool) {
+	const minimizeAttempts = 3
+	job.p, job.call = prog.Minimize(job.p, job.call, false,
+		func(p1 *prog.Prog, call1 int) bool {
+			if stop {
+				return false
+			}
+			for i := 0; i < minimizeAttempts; i++ {
+				result := fuzzer.execWait(job, &Request{
+					Prog:       p1,
+					NeedSignal: true,
+					stat:       statMinimize,
+				})
+				if result.Stop {
+					stop = true
+					return false
+				}
+				info := result.Info
+				if !reexecutionSuccess(info, &job.info, call1) {
+					// The call was not executed or failed.
+					continue
+				}
+				thisSignal, _ := getSignalAndCover(p1, info, call1)
+				if newSignal.Intersection(thisSignal).Len() == newSignal.Len() {
+					return true
+				}
+			}
+			return false
+		})
+	return stop
+}
+
+func reexecutionSuccess(info *ipc.ProgInfo, oldInfo *ipc.CallInfo, call int) bool {
+	if info == nil || len(info.Calls) == 0 {
+		return false
+	}
+	if call != -1 {
+		// Don't minimize calls from successful to unsuccessful.
+		// Successful calls are much more valuable.
+		if oldInfo.Errno == 0 && info.Calls[call].Errno != 0 {
+			return false
+		}
+		return len(info.Calls[call].Signal) != 0
+	}
+	return len(info.Extra.Signal) != 0
+}
+
+func getSignalAndCover(p *prog.Prog, info *ipc.ProgInfo, call int) (signal.Signal, []uint32) {
+	inf := &info.Extra
+	if call != -1 {
+		inf = &info.Calls[call]
+	}
+	return signal.FromRaw(inf.Signal, signalPrio(p, inf, call)), inf.Cover
+}
+
+type smashJob struct {
+	p     *prog.Prog
+	call  int
+	short bool
+}
+
+func (job *smashJob) priority() priority {
+	return smashJobPrio
+}
+
+func (job *smashJob) run(fuzzer *Fuzzer) {
+	fuzzer.Logf(2, "smashing the following program (call=%d, short=%v):\n%s",
+		job.call, job.short, string(job.p.Serialize()))
+	if !job.short {
+		job.scheduleSubjobs(fuzzer)
```

Comment:
I would run these directly here. Synchronous is always better than asynchronous.
I don't think we need to parallelize this. If we have only 1 smash job and nothing else, it's OK to smash sequentially.
IIRC you noticed that all procs can run very similar programs and that's not what we want. In this sense it's even better to not parallelize a single smash.

Reply:
I've squashed collide into smash and made smash do fault injection. Now it only spawns hints.